traversal/expression/Operation.java

```diff
@@ -0,0 +1,293 @@
+        register(new Triple<>("-", Encoding.ValueType.LONG, Encoding.ValueType.LONG), new LongOp.Minus(null, null));
+        register(new Triple<>("*", Encoding.ValueType.LONG, Encoding.ValueType.LONG), new LongOp.Times(null, null));
+        register(new Triple<>("/", Encoding.ValueType.LONG, Encoding.ValueType.LONG), new Cast.CastBoth(new DoubleOp.Div(null, null)));
+        register(new Triple<>("^", Encoding.ValueType.LONG, Encoding.ValueType.LONG), new Unimplemented());
+
+        register(new Triple<>("+", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG), new Cast.CastFirst(new DoubleOp.Plus(null, null)));
+        register(new Triple<>("-", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG), new Cast.CastFirst(new DoubleOp.Minus(null, null)));
+        register(new Triple<>("*", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG), new Cast.CastFirst(new DoubleOp.Times(null, null)));
+        register(new Triple<>("/", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG), new Cast.CastFirst(new DoubleOp.Div(null, null)));
+        register(new Triple<>("^", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG), new Unimplemented());
+
+        register(new Triple<>("+", Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE), new Cast.CastSecond(new DoubleOp.Plus(null, null)));
+        register(new Triple<>("-", Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE), new Cast.CastSecond(new DoubleOp.Minus(null, null)));
+        register(new Triple<>("*", Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE), new Cast.CastSecond(new DoubleOp.Times(null, null)));
+        register(new Triple<>("/", Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE), new Cast.CastSecond(new DoubleOp.Div(null, null)));
+        register(new Triple<>("^", Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE), new Unimplemented());
+    }
+
+    private static void register(Triple<String, Encoding.ValueType<?>, Encoding.ValueType<?>> signature,
+                                 ExpressionFactory.OperationFactory<?, ?> factory) {
+        if (null != registry.put(signature, factory)) throw TypeDBException.of(DUPLICATE_EXPRESSION_DEFINITION, signature);
+    }
+
+    public static ExpressionFactory.OperationFactory resolveOperation(String symbol, Encoding.ValueType<?> arg1Type, Encoding.ValueType<?> arg2Type) {
+        return registry.get(new Triple<>(symbol, arg1Type, arg2Type));
+    }
+
+    protected final Expression<T1> x1;
+    protected final Expression<T2> x2;
+
+    public Operation(Expression<T1> x1, Expression<T2> x2) {
+        this.x1 = x1;
+        this.x2 = x2;
+    }
+
+    public static abstract class DoubleOp extends Operation<java.lang.Double, java.lang.Double> implements Expression.Double {
+        public DoubleOp(Expression<java.lang.Double> x1, Expression<java.lang.Double> x2) {
+            super(x1,x2);
+        }
+
+        public static class Plus extends DoubleOp {
```

Comment:
Invert this - group by Operation name first, then return type - we think about what plus is defined for rather than what can return a double